packages/xo-server/src/xo-mixins/proxies.js

```diff
@@ -60,48 +69,168 @@ export default class Proxy {
     return this._db.remove(id)
   }
 
-  async getProxy(id) {
+  async destroyProxy(id) {
+    const { vmUuid } = await this._getProxy(id)
+    if (vmUuid !== undefined) {
+      await this._app.getXapi(vmUuid).deleteVm(vmUuid)
+    }
+    return this.unregisterProxy(id)
+  }
+
+  async _getProxy(id) {
     const proxy = await this._db.first(id).then(extractProperties)
     if (proxy === undefined) {
       throw noSuchObject(id, 'proxy')
     }
     return proxy
   }
 
+  getProxy(id) {
+    return this._getProxy(id).then(omitToken)
+  }
+
   getAllProxies() {
-    return this._db.get()
+    return this._db.get().then(proxies => proxies.map(omitToken))
   }
 
   @synchronizedWrite
-  async updateProxy(id, { address, authenticationToken, name }) {
-    const proxy = await this.getProxy(id)
-    if (address !== undefined && proxy.address !== address) {
-      await this._throwIfRegistered(address)
-    }
+  async updateProxy(id, { address, authenticationToken, name, vmUuid }) {
+    await this._throwIfRegistered(address, vmUuid)
 
-    patch(proxy, { address, authenticationToken, name })
-    return this._db.update(proxy).then(extractProperties)
+    const proxy = await this._getProxy(id)
+    patch(proxy, { address, authenticationToken, name, vmUuid })
+    return this._db
+      .update(proxy)
+      .then(extractProperties)
+      .then(omitToken)
   }
 
-  async callProxyMethod(id, method, params, expectStream = false) {
-    const proxy = await this.getProxy(id)
-
-    const response = await timeout.call(
-      this._app.httpRequest(proxy.address, {
-        body: format.request(0, method, params),
-        headers: {
-          'Content-Type': 'application/json',
-          Cookie: cookie.serialize(
-            'authenticationToken',
-            proxy.authenticationToken
-          ),
-        },
-        method: 'POST',
-        pathname: '/api/v1',
+  async upgradeProxyAppliance(id) {
+    const { vmUuid } = await this._getProxy(id)
+    const xapi = this._app.getXapi(vmUuid)
+    await xapi.getObject(vmUuid).update_xenstore_data({
+      'vm-data/xoa-updater-channel': JSON.stringify(this._xoProxyConf.channel),
+    })
+
+    return xapi.rebootVm(vmUuid, { hard: true })
```

Comment:
Should we try and remove the xenstore entry if the reboot fails?

Reply:
The user can retry manually. The remove is handled by the updater.

Reply:
The user can retry what? To reboot the VM? But what if the reboot fails here and then the user reboots the VM manually a week later? Will they expect the proxy to be updated?